modules/javafx.base/src/main/java/javafx/beans/value/LazyObjectBinding.java

```diff
@@ -0,0 +1,119 @@
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javafx.beans.value;
+
+import com.sun.javafx.binding.Subscription;
+
+import javafx.beans.InvalidationListener;
+import javafx.beans.binding.ObjectBinding;
+
+/**
+ * Extends {@link ObjectBinding} with the ability to lazily register and eagerly unregister listeners on its
+ * dependencies.
+ *
+ * @param <T> the type of the wrapped {@code Object}
+ */
+abstract class LazyObjectBinding<T> extends ObjectBinding<T> {
+
+    private Subscription subscription;
+    private boolean wasObserved;
+
+    @Override
+    public void addListener(ChangeListener<? super T> listener) {
+        super.addListener(listener);
+
+        updateSubcriptionAfterAdd();
+    }
+
+    @Override
+    public void removeListener(ChangeListener<? super T> listener) {
+        super.removeListener(listener);
+
+        updateSubcriptionAfterRemove();
+    }
+
+    @Override
+    public void addListener(InvalidationListener listener) {
+        super.addListener(listener);
+
+        updateSubcriptionAfterAdd();
+    }
+
+    @Override
+    public void removeListener(InvalidationListener listener) {
+        super.removeListener(listener);
+
+        updateSubcriptionAfterRemove();
+    }
+
+    @Override
+    protected boolean allowValidation() {
+        return isObserved();
+    }
+
+    /**
+     * Called after a listener was added to start observing inputs if they're not observed already.
+     */
+    private void updateSubcriptionAfterAdd() {
+        if (!wasObserved) { // was first observer registered?
+            subscription = observeInputs(); // start observing source
+
+            /*
+             * Although the act of registering a listener already attempts to make
+             * this binding valid, allowValidation won't allow it as the binding is
+             * not observed yet. This is because isObserved will not yet return true
+             * when the process of registering the listener hasn't completed yet.
+             *
+             * As the binding must be valid after it becomes observed the first time
+             * 'get' is called again.
```

Comment:
Maybe add an `@link` to `ExpressionHelper` where the validation happens before listener registration, for clarity.

Reply:
I've added this additional text:

             * See com.sun.javafx.binding.ExpressionHelper (which is used
             * by ObjectBinding) where it will do a call to ObservableValue#getValue
             * BEFORE adding the actual listener. This results in ObjectBinding#get
             * to be called in which the #allowValidation call will block it from
             * becoming valid as the condition is "isObserved()"; this is technically
             * correct as the listener wasn't added yet, but means we must call
             * #get again to make this binding valid.

This might be something we can fix in future work, to avoid this double call to `get`.